code/eyetracking/eyetrackingrun.py

```diff
@@ -0,0 +1,746 @@
+                & (self.samples["time"] <= saccade_event["end"]),
+                "saccade",
+            ] = 1
+
+            if saccade_event["blink"] == 1:
+                self.samples.loc[
+                    (self.samples["time"] >= saccade_event["start"])
+                    & (self.samples["time"] <= saccade_event["end"]),
+                    "blink",
+                ] = 1
+
+        return self.samples
+
+    def find_timestamp_message(self) -> Optional[int]:
+        """
+        Finds the first row in the DataFrame containing the specified message string and returns the 'trialid_time' as an integer.
+
+        Returns:
+            Optional[int]: The 'trialid_time' as an integer if the message is found; None if the message is not found.
+        """
+        message_row = self.messages[
+            self.messages["trialid "].str.contains(
+                self.message_first_trigger, case=False, regex=True
+            )
+        ].head(1)
+        if not message_row.empty:
+            return int(message_row["trialid_time"].iloc[0])
+
+    def extract_calibration(
+        self,
+    ) -> Tuple[
+        int,
+        Optional[str],
+        Optional[float],
+        Optional[float],
+        Optional[List[List[Union[int, int]]]],
+    ]:
+        """
+        Extracts calibration information from the DataFrame of messages.
+
+        Returns:
+        Tuple[int, Optional[str], Optional[float], Optional[float], Optional[List[List[Union[int, int]]]]]:
+            A tuple containing:
+            - Calibration count (int),
+            - Calibration type (str or None),
+            - Average calibration error (float or None),
+            - Maximum calibration error (float or None),
+            - Calibration position (list of lists of integers or None).
+        """
+
+        row_error_value = self.messages[
+            self.messages["trialid "].str.contains("ERROR", case=False, regex=True)
+        ].head(1)
+
+        if row_error_value.empty:
+            calibration_count = 0
+            print("No calibration information found.")
+            return calibration_count, None, None, None, None
+        else:
+            calibration_count = 1
+            error_message = row_error_value["trialid "].iloc[0]
+
+            matches = re.findall(r"([-+]?\d*\.\d+|\d+)", error_message)
+            average_calibration_error, max_calibration_error = (
+                map(float, matches[1:3]) if len(matches) >= 2 else (None, None)
+            )
+
+            print("Calibration Count:", calibration_count)
+            print("Average Calibration Error:", average_calibration_error)
+            print("Maximum Calibration Error:", max_calibration_error)
+
+            if "HV9" in error_message:
+                calibration_type = "HV9"
+                calibration_position = self.get_calibration_positions(calibration_type)
+            else:
+                calibration_type = None
+                calibration_position = None
```

Comment:
We can easily support most of the possibilities with about the same number of lines.
```suggestion
            calibration_type = None
            calibration_position = None
            if (points := re.findall(r"HV(\d{1,2})", error_message)):
                calibration_type = f"HV{points[0]}"
                calibration_position = EYELINK_CALIBRATION_COORDINATES[:int(points[0])]
```